File: backend/internal/services/embedding_service_test.go
package services_test

import (
	"context"
	"errors"
	"strings"
	"testing"
	"time"

	"nexia-backend/internal/models"
	"nexia-backend/internal/services"

	"github.com/stretchr/testify/require"
	"go.uber.org/mock/gomock"
	"go.uber.org/zap"
	"gorm.io/gorm"
)

type embeddingFakeProfileRepo struct {
	profile *models.Profile
	loadErr error
}

func (f *embeddingFakeProfileRepo) Create(_ *models.Profile) error { return nil }
func (f *embeddingFakeProfileRepo) FindByID(_ uint64, _ uint64) (*models.Profile, error) {
	return nil, nil
}
func (f *embeddingFakeProfileRepo) FindAll(_ int, _ int, _ string, _ string, _ uint64) ([]models.Profile, int64, error) {
	return nil, 0, nil
}
func (f *embeddingFakeProfileRepo) Update(_ *models.Profile) error  { return nil }
func (f *embeddingFakeProfileRepo) Delete(_ uint64, _ uint64) error { return nil }
func (f *embeddingFakeProfileRepo) LoadForEmbedding(_ context.Context, _ uint) (*models.Profile, error) {
	return f.profile, f.loadErr
}

func newEmbeddingSvc(profiles *embeddingFakeProfileRepo, gen services.EmbeddingGenerator, repo services.EmbeddingRepository) *services.EmbeddingService {
	return services.NewEmbeddingService(profiles, gen, repo, zap.NewNop())
}

func TestEmbedProfileSuccess(t *testing.T) {
	ctrl := gomock.NewController(t)
	gen := NewMockEmbeddingGenerator(ctrl)
	repo := NewMockEmbeddingRepository(ctrl)

	birthday := models.Date(time.Date(2000, time.March, 21, 0, 0, 0, 0, time.UTC))
	profile := &models.Profile{
		ID:               7,
		UserID:           9,
		FullName:         "Alice Example",
		Bio:              "A close friend",
		Profession:       "Engineer",
		RelationshipType: "Friend",
		Birthday:         &birthday,
		LongTermGoals:    "Build a startup",
		MusicPreference:  "Rock",
		FavoriteMovie:    "Inception",
		FavoriteBook:     "Dune",
		FavoriteMemory:   "Road trip",
		Notes:            "Met at a hackathon",
		Tags:             []models.Tag{{Tag: "travel"}},
		PoliticalViews:   []models.PoliticalView{{View: "Moderate"}},
		FoodRestrictions: []models.FoodRestriction{{Restriction: "None"}},
		MovieGenres:      []models.MovieGenre{{Genre: "Sci-Fi"}},
		BookGenres:       []models.BookGenre{{Genre: "Fantasy"}},
		HangoutPlaces:    []models.HangoutPlace{{Place: "Cafe"}},
		Quotes:           []models.Quote{{Quote: "Keep going"}},
		TopSongs:         []models.TopSong{{Name: "Song A", Artist: "Artist A"}},
		AssociatedSong:   &models.AssociatedSong{Name: "Song B", Artist: "Artist B"},
	}

	var generatedText string
	embedding := []float32{0.1, 0.2}
	gen.EXPECT().
		GenerateEmbedding(gomock.Any(), gomock.Any()).
		DoAndReturn(func(_ context.Context, text string) ([]float32, error) {
			generatedText = text
			return embedding, nil
		})
	repo.EXPECT().UpsertProfile(gomock.Any(), profile, embedding).Return(nil)

	svc := newEmbeddingSvc(&embeddingFakeProfileRepo{profile: profile}, gen, repo)
	require.NoError(t, svc.EmbedProfile(context.Background(), 7))

	for _, want := range []string{
		"Profile of Alice Example",
		"Political Views: Moderate",
		"Food Restrictions: None",
		"Favorite Movie Genres: Sci-Fi",
		"Favorite Book Genres: Fantasy",
		"Favorite Hangout Places: Cafe",
		"Associated Song: Song B by Artist B",
		"Top Songs: Song A by Artist A",
		`"Keep going"`,
		"Notes: Met at a hackathon",
	} {
		require.True(t, strings.Contains(generatedText, want))
	}
}

func TestEmbedProfileNotFound(t *testing.T) {
	ctrl := gomock.NewController(t)
	svc := newEmbeddingSvc(&embeddingFakeProfileRepo{loadErr: gorm.ErrRecordNotFound}, NewMockEmbeddingGenerator(ctrl), NewMockEmbeddingRepository(ctrl))

	err := svc.EmbedProfile(context.Background(), 1)
	require.ErrorIs(t, err, services.ErrNotFound)
}

func TestEmbedProfileLoadError(t *testing.T) {
	ctrl := gomock.NewController(t)
	svc := newEmbeddingSvc(&embeddingFakeProfileRepo{loadErr: errors.New("db connection lost")}, NewMockEmbeddingGenerator(ctrl), NewMockEmbeddingRepository(ctrl))

	err := svc.EmbedProfile(context.Background(), 1)
	require.Error(t, err)
	require.False(t, errors.Is(err, services.ErrNotFound))
}

func TestEmbedProfileGenerationError(t *testing.T) {
	ctrl := gomock.NewController(t)
	gen := NewMockEmbeddingGenerator(ctrl)
	repo := NewMockEmbeddingRepository(ctrl)

	gen.EXPECT().GenerateEmbedding(gomock.Any(), gomock.Any()).Return(nil, errors.New("gemini unavailable"))

	svc := newEmbeddingSvc(&embeddingFakeProfileRepo{profile: &models.Profile{ID: 1, FullName: "Bob"}}, gen, repo)
	require.Error(t, svc.EmbedProfile(context.Background(), 1))
}

func TestEmbedProfileUpsertError(t *testing.T) {
	ctrl := gomock.NewController(t)
	gen := NewMockEmbeddingGenerator(ctrl)
	repo := NewMockEmbeddingRepository(ctrl)

	profile := &models.Profile{ID: 1, FullName: "Bob"}
	embedding := []float32{0.1, 0.2}
	gen.EXPECT().GenerateEmbedding(gomock.Any(), gomock.Any()).Return(embedding, nil)
	repo.EXPECT().UpsertProfile(gomock.Any(), profile, embedding).Return(errors.New("pgvector unavailable"))

	svc := newEmbeddingSvc(&embeddingFakeProfileRepo{profile: profile}, gen, repo)
	require.Error(t, svc.EmbedProfile(context.Background(), 1))
}

func TestDeleteEmbeddingSuccess(t *testing.T) {
	ctrl := gomock.NewController(t)
	repo := NewMockEmbeddingRepository(ctrl)
	repo.EXPECT().DeleteProfile(gomock.Any(), uint64(42)).Return(nil)

	svc := newEmbeddingSvc(&embeddingFakeProfileRepo{}, NewMockEmbeddingGenerator(ctrl), repo)
	require.NoError(t, svc.DeleteEmbedding(context.Background(), 42))
}

func TestDeleteEmbeddingError(t *testing.T) {
	ctrl := gomock.NewController(t)
	repo := NewMockEmbeddingRepository(ctrl)
	repo.EXPECT().DeleteProfile(gomock.Any(), uint64(1)).Return(errors.New("store unavailable"))

	svc := newEmbeddingSvc(&embeddingFakeProfileRepo{}, NewMockEmbeddingGenerator(ctrl), repo)
	require.Error(t, svc.DeleteEmbedding(context.Background(), 1))
}

File: apps/api/src/services/embedding-service.test.ts
import { describe, test, expect } from "bun:test";
import type { ProfileOutput } from "@nexia/shared";
import {
  EmbeddingService,
  type EmbeddingGenerator,
  type EmbeddingStorage,
  type ProfileLoader,
} from "../services/embedding-service";
import { ServiceError, ErrorKind } from "../services/errors";
import pino from "pino";
const logger = pino({ level: "silent" });

class FakeEmbeddingGenerator implements EmbeddingGenerator {
  generatedText = "";
  err: Error | null = null;
  async generateEmbedding(text: string) {
    if (this.err) throw this.err;
    this.generatedText = text;
    return new Array<number>(3072).fill(0.001) as number[];
  }
}

class FakeEmbeddingStorage implements EmbeddingStorage {
  upserted: Array<{
    profileId: number;
    userId: number;
    embedding: number[];
    payload: Record<string, unknown>;
  }> = [];
  deletedProfiles: number[] = [];
  upsertErr: Error | null = null;
  deleteErr: Error | null = null;

  async upsertProfile(
    profileId: number,
    userId: number,
    embedding: number[],
    payload: Record<string, unknown>
  ) {
    if (this.upsertErr) throw this.upsertErr;
    this.upserted.push({ profileId, userId, embedding, payload });
  }
  async deleteProfile(profileId: number) {
    if (this.deleteErr) throw this.deleteErr;
    this.deletedProfiles.push(profileId);
  }
}

describe("EmbeddingService", () => {
  function fullProfileRepo(): ProfileLoader {
    return {
      async loadForEmbedding(_profileId: number): Promise<ProfileOutput> {
        return {
          id: 7,
          user_id: 42,
          full_name: "Alice",
          relationship_type: "Friend",
          bio: "A great friend",
          profession: "Engineer",
          long_term_goals: "Travel",
          birthday: "2001-03-22",
          zodiac_sign: "Aries",
          music_preference: "Pop",
          favorite_movie: "Inception",
          favorite_book: "Dune",
          notes: "Awesome",
          created_at: "2026-01-01T00:00:00.000Z",
          updated_at: "2026-01-01T00:00:00.000Z",
          tags: [
            { id: 1, profile_id: 7, tag: "kind" },
            { id: 2, profile_id: 7, tag: "funny" },
          ],
          political_views: [{ id: 1, profile_id: 7, view: "moderate" }],
          food_restrictions: [{ id: 1, profile_id: 7, restriction: "vegetarian" }],
          movie_genres: [{ id: 1, profile_id: 7, genre: "sci-fi" }],
          book_genres: [{ id: 1, profile_id: 7, genre: "fiction" }],
          hangout_places: [{ id: 1, profile_id: 7, place: "coffee shop" }],
          quotes: [{ id: 1, profile_id: 7, quote: "Be yourself" }],
          favorite_memories: [{ id: 1, profile_id: 7, memory: "Road trip" }],
          top_songs: [{ id: 1, profile_id: 7, name: "Song 1", artist: "Artist 1" }],
          associated_song: { profile_id: 7, name: "Anthem", artist: "Band" },
        };
      },
    };
  }

  function emptyProfileRepo(): ProfileLoader {
    return {
      async loadForEmbedding(_profileId: number) {
        return null;
      },
    };
  }

  test("embed profile success", async () => {
    const profiles = fullProfileRepo();
    const generator = new FakeEmbeddingGenerator();
    const repo = new FakeEmbeddingStorage();
    const svc = new EmbeddingService(profiles, generator, repo, logger);

    await svc.embedProfile(7);
    expect(repo.upserted.length).toBe(1);
    expect(repo.upserted[0]!.profileId).toBe(7);
    expect(repo.upserted[0]!.userId).toBe(42);
    expect(generator.generatedText).toContain("Profile of Alice");
    expect(generator.generatedText).toContain("March 22, 2001");
    expect(generator.generatedText).toContain('"Be yourself"');
    expect(generator.generatedText).toContain("Favorite Memories: Road trip");
  });

  test("embed profile not found", async () => {
    const profiles = emptyProfileRepo();
    const svc = new EmbeddingService(
      profiles,
      new FakeEmbeddingGenerator(),
      new FakeEmbeddingStorage(),
      logger
    );

    try {
      await svc.embedProfile(7);
      expect.unreachable();
    } catch (err) {
      expect(err instanceof ServiceError).toBe(true);
      expect((err as ServiceError).kind).toBe(ErrorKind.NotFound);
    }
  });

  test("embed profile generation error", async () => {
    const profiles = fullProfileRepo();
    const generator = new FakeEmbeddingGenerator();
    generator.err = new Error("api error");
    const svc = new EmbeddingService(profiles, generator, new FakeEmbeddingStorage(), logger);

    try {
      await svc.embedProfile(7);
      expect.unreachable();
    } catch (err) {
      expect((err as Error).message).toContain("gemini embedding failed");
    }
  });

  test("embed profile upsert error", async () => {
    const profiles = fullProfileRepo();
    const repo = new FakeEmbeddingStorage();
    repo.upsertErr = new Error("db error");
    const svc = new EmbeddingService(profiles, new FakeEmbeddingGenerator(), repo, logger);

    try {
      await svc.embedProfile(7);
      expect.unreachable();
    } catch (err) {
      expect((err as Error).message).toContain("pgvector upsert failed");
    }
  });

  test("delete embedding success", async () => {
    const repo = new FakeEmbeddingStorage();
    const svc = new EmbeddingService(fullProfileRepo(), new FakeEmbeddingGenerator(), repo, logger);

    await svc.deleteEmbedding(7);
    expect(repo.deletedProfiles).toEqual([7]);
  });

  test("delete embedding error", async () => {
    const repo = new FakeEmbeddingStorage();
    repo.deleteErr = new Error("db error");
    const svc = new EmbeddingService(fullProfileRepo(), new FakeEmbeddingGenerator(), repo, logger);

    try {
      await svc.deleteEmbedding(7);
      expect.unreachable();
    } catch (err) {
      expect((err as Error).message).toBe("db error");
    }
  });
});
